pkg/controllers/user/nodesyncer/cordonfieldssyncer.go

```diff
@@ -75,21 +75,26 @@ func (d *nodeDrain) drainNode(key string, obj *v3.Node) (runtime.Object, error)
 		return nil, nil
 	}
 
+	nodeMapLock.Lock()
+	defer nodeMapLock.Unlock()
+
 	if obj.Spec.DesiredNodeUnschedulable == "drain" {
-		nodeMapLock.Lock()
 		if _, ok := d.nodesToContext[obj.Name]; ok {
-			nodeMapLock.Unlock()
 			return nil, nil
 		}
+		node, err := nodehelper.GetNodeForMachine(obj, d.nodeLister)
+		if err != nil {
+			return nil, err
+		}
+		if node == nil {
+			return nil, fmt.Errorf("nodeDrain: error finding node [%s]", obj.Spec.RequestedHostname)
+		}
 		ctx, cancel := context.WithCancel(d.ctx)
 		d.nodesToContext[obj.Name] = cancel
```

Comment:
i know it's cleaner with the defer but wrapping the lock around the if/else means it locks the lister call which you dont need to do since the lock was only for protecting this map. i think you should go back to it being inside the if/else and move locking in the if under the lister.

Reply:
yep you're right, was trying to make it cleaner. but lister blocking might be an issue when multiple nodes are drained simultaneously if it takes a long time. updated!

will create a backport to update 2.4 once approved